modules/afprog/afprog.c

```diff
@@ -91,10 +91,72 @@ _exec_program(const gchar *cmdline)
   execl("/bin/sh", "/bin/sh", "-c", cmdline, NULL);
 }
 
+static void
+_closing_all_fd_expect(int ignore[], int ignore_len)
+{
+  for (int i = 10000; i > 5; --i)
+    {
+      gboolean skip = FALSE;
+      for (int j = 0; j < ignore_len; ++j)
+        if (ignore[j] == i) skip = TRUE;
+
+      if (!skip)
+        close(i);
+    }
+}
+
+static const gchar sync_msg[] = "sync";
+static const gint  sync_msg_len = sizeof(sync_msg)/sizeof(sync_msg[0])-1;
+
+static void
+_signal_parent_fd_closed(int pipe[2])
+{
+  gint written = 0;
+  do
+    {
+      gint result = write(pipe[1], sync_msg+written, sync_msg_len-written);
+      if (result < 0)
+        {
+          msg_warning("afprog(): sending sync to parent failed", evt_tag_error(EVT_TAG_OSERROR));
+          return;
+        }
+      written += result;
+    }
+  while (written < sync_msg_len);
+}
```

Comment:
this can be simplified a bit, as we don't need a message. it is enough to just close the write sides of a pipe and the reader would get a zero sized read.

so parent does a blocking read(), and once it returns indicating an EOF, that child is guaranteed to have closed the fd.
no need to read/write actual data.

Reply:
yeah, the first version was like that but it did not worked for me, there was an error and such it woun't read EOF. now fixed, and simple version also works.